File: src/first_example.R
#
# Copyright (c) 2022 geneXplain GmbH, Wolfenbüttel, Germany
#
# Author: Philip Stegmaier, philip.stegmaier@genexplain.com
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY 
# OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT
# LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
# FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO
# EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE 
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, 
# WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE
# OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#

library(geneXplainR)

# Username (email) and password need to be replaced with valid credentials.
#
gx.login("https://platform.genexplain.com", "someuser@email.io", "12345")

# The gx.ls function returns a listing of the specified workspace folder.
#
gx.ls("data/Projects")

# The project name has to be unique on the connected platform server.
#
apiProjectName <- "api2022_tutorial"
apiProjectPath <- paste0("data/Projects/", apiProjectName)

# The gx.createProject function returns an error if the project already
# exists.
#
gx.createProject(apiProjectName, description = "API 2022 tutorial project")

funclassFolder <- "first_example_functional_class"
folderPath     <- paste0(apiProjectPath, "Data/", funclassFolder)

# The gx.createFolder function creates the folder if it does not
# already exist.
#
gx.createFolder(paste0(apiProjectPath, "Data"), funclassFolder)

species <- "Human (Homo sapiens)"
dataName <- "funclass_genes_covid-19"

# The gx.importers function returns a list of available importers
# for different data types.
#
gx.importers()

# The available parameters for a specific importer and a destination
# folder can be interrogated using the gx.import.parameters function.
#
gx.import.parameters(folderPath, "Tabular (*.txt, *.xls, *.tab, etc.)")

# The geneXplainR library provides a specialized function to import tabular
# data files.
#
gx.importTable("../data/functional_classification_genes_GSE156063.tsv", folderPath,
               dataName, columnForID = "ID", 
               tableType = "Genes: Ensembl", species = species)
               
# This is the platform workspace path of the imported data.
#
funclassGenePath <- paste0(folderPath, "/", dataName)

funclassResultPath <- paste0(funclassGenePath," GO")

# The gx.analysis.parameters function shows input option names and 
# descriptions for a specified platform tool.
#
gx.analyis.parameters("Functional classification")

# The gx.analysis function submits analysis tasks to the connected
# platform server.
#
gx.analysis("Functional classification", list(sourcePath = funclassGenePath,
                                              species    = species,
                                              bioHub     = "Full gene ontology classification",
                                              minHits    = 1,
                                              pvalueThreshold = 1,
                                              outputTable = funclassResultPath))

# The gx.get function loads data tables from the platform workspace into
# a data frame.
#
funclassResult <- gx.get(funclassResultPath)
head(funclassResult)

File: src/FirstExample.java
/**
 * Copyright (C) 2022 geneXplain GmbH, Wolfenbuettel, Germany
 *
 * Author: Philip Stegmaier
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY 
 * OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT
 * LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO
 * EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE 
 * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, 
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE
 * OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 *
 */

import com.eclipsesource.json.JsonValue;
import com.eclipsesource.json.JsonObject;
import com.eclipsesource.json.PrettyPrint;

import com.genexplain.api.core.GxHttpClient;
import com.genexplain.api.core.GxHttpClientImpl;
import com.genexplain.api.core.GxHttpConnection;
import com.genexplain.api.core.GxHttpConnectionImpl;

import java.io.StringWriter;
import java.io.Writer;

public class FirstExample {
    public static void main(String[] args) throws Exception {
    
        // The GxHttpConnectionImpl holds the connection to the specified
        // platform server. Username and password need to correspond to a
        // valid account on that server.
        //
        GxHttpConnectionImpl con = new GxHttpConnectionImpl();
        con.setServer("https://platform.genexplain.com");
        con.setUsername("someuser@email.io");
        con.setPassword("12345");
        con.setVerbose(true);
        con.login();
        
        // The connection is given to a client object. The client is the 
        // main component to interact with the connected platform server.
        //
        GxHttpClientImpl client = new GxHttpClientImpl();
        client.setConnection(con);
        
        // The createProject method creates a new project in the user
        // workspace. The project name must be new and unique on the 
        // platform instance. The platform path of the new project is 
        // "data/Projects/<project name>".
        //
        Map<String, String> projectParams = new HashMap<>();
        projectParams.put("user", "someuser@email.io");
        projectParams.put("pass", "12345");
        projectParams.put("project", "api2022_tutorial");
        projectParams.put("description", "API 2022 tutorial project");
        client.createProject(projectParams)
        
        // The createFolder function creates the folder if it does not 
        // already exist.
        //
        client.createFolder("data/Projects/api2022_tutorial/Data", "first_example_functional_class");
        
        // The API provides for a specialized method to import tabular
        // data.
        //
        client.importTable("../data/functional_classification_genes_GSE156063.tsv",
                           "data/Projects/api2022_tutorial/Data/first_example_functional_class",
                           "funclass_genes_covid-19",
                           false, 
                           GxHttpClient.ColumnDelimiter.Tab, 
                           1, 2, "", "ID", false, "Genes: Ensembl", "Human (Homo sapiens)");
        
        // After data import it is often advisable to let background 
        // processes on the server finish their work, before proceeding 
        // with analyzing the uploaded data.
        //
        Thread.sleep(1000);
        
        // The analyze function submits analysis tasks to the connected 
        // platform server.
        //
        JsonObject params = new JsonObject()
            .add("sourcePath", "data/Projects/api2022_tutorial/Data/first_example_functional_class/funclass_genes_covid-19")
            .add("species", "Human (Homo sapiens)")
            .add("bioHub", "Full gene ontology classification")
            .add("minHits", 1)
            .add("pvalueThreshold", 1)
            .add("outputTable", "data/Projects/api2022_tutorial/Data/first_example_functional_class/funclass_genes_covid-19 GO");
        client.analyze("Functional classification", params, false, true, true);
        
        con.logout();
    }
}
